interface.go

```diff
@@ -19,100 +19,46 @@ package reuseport
 
 import (
 	"context"
-	"errors"
 	"net"
-	"syscall"
-	"time"
+
+	"github.com/pkg/errors"
 )
 
-// Available returns whether or not SO_REUSEPORT is available in the OS.
-// It does so by attepting to open a tcp listener, setting the option, and
-// checking ENOPROTOOPT on error. After checking, the decision is cached
-// for the rest of the process run.
+// Available returns whether or not SO_REUSEPORT or equivalent behaviour is
+// available in the OS.
 func Available() bool {
-	return available()
+	return true
 }
 
-// ErrUnsuportedProtocol signals that the protocol is not currently
-// supported by this package. This package currently only supports TCP.
-var ErrUnsupportedProtocol = errors.New("protocol not yet supported")
-
-// ErrReuseFailed is returned if a reuse attempt was unsuccessful.
-var ErrReuseFailed = errors.New("reuse failed")
-
-// ErrDialSelf is returned if we connect to our own source address.
-var ErrDialSelf = errors.New("dialed our own socket")
+var listenConfig = net.ListenConfig{
+	Control: Control,
+}
 
 // Listen listens at the given network and address. see net.Listen
 // Returns a net.Listener created from a file discriptor for a socket
 // with SO_REUSEPORT and SO_REUSEADDR option set.
 func Listen(network, address string) (net.Listener, error) {
-	if !available() {
-		return nil, syscall.ENOPROTOOPT
-	}
-
-	return listenStream(network, address)
+	return listenConfig.Listen(context.Background(), network, address)
 }
 
 // ListenPacket listens at the given network and address. see net.ListenPacket
 // Returns a net.Listener created from a file discriptor for a socket
 // with SO_REUSEPORT and SO_REUSEADDR option set.
 func ListenPacket(network, address string) (net.PacketConn, error) {
-	if !available() {
-		return nil, syscall.ENOPROTOOPT
-	}
-
-	return listenPacket(network, address)
+	return listenConfig.ListenPacket(context.Background(), network, address)
 }
 
 // Dial dials the given network and address. see net.Dialer.Dial
 // Returns a net.Conn created from a file discriptor for a socket
 // with SO_REUSEPORT and SO_REUSEADDR option set.
 func Dial(network, laddr, raddr string) (net.Conn, error) {
-	if !available() {
-		return nil, syscall.ENOPROTOOPT
+	nla, err := ResolveAddr(network, laddr)
+	if err != nil {
+		return nil, errors.Wrap(err, "resolving local addr")
 	}
-
-	var d Dialer
-	if laddr != "" {
-		netladdr, err := ResolveAddr(network, laddr)
-		if err != nil {
-			return nil, err
-		}
-		d.D.LocalAddr = netladdr
+	d := net.Dialer{
+		Control:   Control,
+		LocalAddr: nla,
 	}
-
 	return d.Dial(network, raddr)
 }
-
-// Dialer is used to specify the Dial options, much like net.Dialer.
-// We simply wrap a net.Dialer.
-type Dialer struct {
```

Comment:
We should probably keep and deprecate this. It's a public interface and it's actually being used externally (unlike the poll package). We don't *have* to (we can bump to 0.2.0 and make a breaking change) but I'd rather keep it around (deprecated) for a while.

To be completely correct, we'll probably need to (a) copy the `net.Dialer` before replacing the `Control` function and (b) chain the control functions.

Reply:
We have one user here: https://github.com/sonm-io/core/blob/c40eaa1a84fdc6faa9d837e9df1b4c16065aa194/insonmnia/npp/net.go#L26. However, they're using go-mod.

See: https://godoc.org/github.com/libp2p/go-reuseport?importers

---

But I'm not going to object as long as we have a patch ready for go-reuseport-transport.

Reply:
This is breaking our CI builds downstream in github.com/prysmaticlabs/prysm

Reply:
@prestonvanloon can you link to log for the failed job?

Reply:
https://travis-ci.org/prysmaticlabs/prysm/jobs/475121335

Reply:
We are fetching all dependencies from HEAD right now, but this is something we are aiming to move away from as we are frequently broken by upstream changes.

Reply:
Consider using gx, or go.mod which I've included in libp2p/go-reuseport-transport#6.

Reply:
Note: to be clear, we support using libp2p directly from master with `go get` for now.

---

I should have been clear and said "merge at the same time" instead of "have the patch ready". We should also *document* what dependency systems we support somewhere.